x/lockup/keeper/lock.go

```diff
@@ -849,3 +849,70 @@ func (k Keeper) getCoinsFromLocks(locks []types.PeriodLock) sdk.Coins {
 	}
 	return coins
 }
+
+// RebondTokens rebonds some amount or all coins of a lock with id lockID only if it is already unbonding
+func (k Keeper) RebondTokens(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coins sdk.Coins) error {
+	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return err
+	}
+
+	if lock.Owner != owner.String() {
+		return fmt.Errorf("lock %d is not owned by %s", lockID, owner)
+	}
+
+	if !lock.IsUnlocking() {
+		return errorsmod.Wrap(types.ErrLockNotUnlocking, fmt.Sprintf("lock %d is not unlocking, rebonding only possible in unlocking stage", lockID))
+	}
+
+	// If all checks pass, we can rebond the tokens
+	var rebondedLock types.PeriodLock
+	// partial re-bonding
+	if coins != nil && !coins.IsEqual(lock.Coins) {
+		// This branch implies we will end up with two locks:
+		// - The first lock will be the original lock with the coins removed, in unlocking state
+		// - The second lock will be the new rebonded lock with the coins added, in not unlocking state
+
+		// remove rebonded lock's coins from original lock
+		lock.Coins = lock.Coins.Sub(coins)
+		err = k.setLock(ctx, *lock)
+		if err != nil {
+			return errorsmod.Wrap(types.ErrFailedToSetLock, fmt.Sprintf("failed to set original lock %d: %s", lock.ID, err.Error()))
+		}
+
+		// create a rebonded lock
+		splitLockID := k.GetLastLockID(ctx) + 1
+		k.SetLastLockID(ctx, splitLockID)
+		rebondedLock = types.NewPeriodLock(splitLockID, owner, owner.String(), lock.Duration, time.Time{}, coins)
```

Comment:
for reviewers: do we need to make it possible for a user to specify `reward_address` when rebonding? I personally think it would be nice to have

Reply:
cc: @mattverse @p0mvn

Reply:
@czarcas7ic @p0mvn what do you guys think on this one?

Reply:
I personally think that we might go with this feature included already. It is already made optional (as long as a person does not specify `amount` as a flag)